Lib/test/test_capi.py

```diff
@@ -620,14 +620,30 @@ def test_heaptype_with_custom_metaclass(self):
 
     def test_pynumber_tobase(self):
         from _testcapi import pynumber_tobase
-        self.assertEqual(pynumber_tobase(123, 2), '0b1111011')
-        self.assertEqual(pynumber_tobase(123, 8), '0o173')
-        self.assertEqual(pynumber_tobase(123, 10), '123')
-        self.assertEqual(pynumber_tobase(123, 16), '0x7b')
-        self.assertEqual(pynumber_tobase(-123, 2), '-0b1111011')
-        self.assertEqual(pynumber_tobase(-123, 8), '-0o173')
-        self.assertEqual(pynumber_tobase(-123, 10), '-123')
-        self.assertEqual(pynumber_tobase(-123, 16), '-0x7b')
+        large_number = 2**64
+        small_number = 123
+        # subclass of int
+        class TrapInt(int):
+            def __index__(self):
+                return int(self)
```

Comment:
Because `TrapInt` is a subclass of `int`, this never actually calls `__index__`. `PyNumber_Index` uses `PyLong_Check` internally and simply returns the input object if it passes that (plus later converting it to an actual `int`). `PyLong_Check(o)` is effectively `isinstance(o, int)`. While it might be worth testing that conversion of subclasses, that's something for the index tests. Here, we need an object that isn't an `int` subclass.

```
>>> class TrapInt(int):
...  def __index__(self):
...   print("index")
...   return int(self)
... 
>>> # Explicitly calling it calls __index__, of course:
>>> TrapInt(1).__index__()
index
1
>>> # But operator.index (= PyNumber_Index) does not:
>>> operator.index(TrapInt(42))
42
```

Reply:
Right. The `TrapInt` class is meaningless for testing `PyNumber_ToBase`, it behaves the same way as `int`. I would like to replace `TrapInt` with the following class.
```python
class IDX(object):
    def __init__(self, val):
        self.val = val
    def __index__(self):
        return self.val
```
Also, I looked at the docs of `PyNumber_ToBase(PyObject *n, int base)`, which states that `If *n* is not a Python int, it is converted with PyNumber_Index first.` Wouldn't it be better to change this sentence to `If *n* is not a Python int or subclass of int, ...`

Reply:
Yes, that class is what I'd do (although the `(object)` bit is unnecessary with Py3; all classes are based on `object`).

Regarding the suggested doc change: I'm not sure that's necessary. 'If *x* is a *type*' should be understood as `isinstance(x, type)`. See for example the docs of `PyLong_AsLong` et al. Functions which expect only *exactly* the specified type mention that explicitly, e.g. `PyLong_CheckExact`.

However, I see that several of the `PyNumber_*` function descriptions use the Python object names (`float` and `int`) rather than the C names (`PyFloatObject` and `PyLongObject`) as is otherwise used in much of the C API documentation. That might be worth considering changing.